pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/rule/security/lib/ElEscapeDetector.java

```diff
@@ -0,0 +1,234 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.vf.rule.security.lib;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.vf.DataType;
+import net.sourceforge.pmd.lang.vf.ast.ASTArguments;
+import net.sourceforge.pmd.lang.vf.ast.ASTDotExpression;
+import net.sourceforge.pmd.lang.vf.ast.ASTElExpression;
+import net.sourceforge.pmd.lang.vf.ast.ASTExpression;
+import net.sourceforge.pmd.lang.vf.ast.ASTIdentifier;
+import net.sourceforge.pmd.lang.vf.ast.ASTNegationExpression;
+import net.sourceforge.pmd.lang.vf.ast.AbstractVFNode;
+import net.sourceforge.pmd.lang.vf.ast.VfTypedNode;
+
+/**
+ * Helps detect visualforce encoding in EL Expressions
+ * (porting over code previously living in VfUnescapeElRule for reusability)
+ */
+
+public final class ElEscapeDetector {
```

Comment:
Class to help with avoiding code duplication. The contents here are identical to what was previously present in `VfUnescapeElRule.java`

Note: I attempted to make the methods static and have a private constructor to comply with PMD's utility class definition. However, these methods in static form created regressions in the functioning of VfUnescapeEl rule. I've now restored them back to instance-level and the regressions went away as well.

Reply:
The regression was from the apex layer that detected types that were safe and did not need encoding inside ElExpressions. It started throwing many false positives on types that we know are type safe. My guess is that there must've been an instance-level interaction which caused a confusion in the type detected.